docs/man/rpmsign.8.md
RPMSIGN(8)

# NAME

rpmsign - RPM Package Signing

# SYNOPSIS

*rpmsign* *--addsign|--resign* [*rpmsign-options*] *PACKAGE_FILE
...*

*rpmsign* *--delsign* *PACKAGE_FILE ...*

*rpmsign* *--delfilesign* *PACKAGE_FILE ...*

## rpmsign-options

\[*--rpmv3*] [*--rpmv4*] [*--fskpath* *KEY*] [*--signfiles*]

# DESCRIPTION

*rpmsign* *--addsign* generates and inserts a new OpenPGP signature
for each *PACKAGE_FILE* given unless a signature with identical
parameters already exists, in which case no action is taken.
Arbitrary number of V6 signatures can be added.

*rpmsign* *--resign* generates and inserts a new OpenPGP signature
for each *PACKAGE_FILE*, replacing any and all previous signatures.

To create a signature rpmsign needs to verify the package's checksum. As a
result V4 packages with MD5/SHA1 checksums cannot be signed in FIPS mode.

*rpmsign* *--delsign* *PACKAGE_FILE ...*

Delete all OpenPGP signatures from each package *PACKAGE_FILE* given.

*rpmsign* *--delfilesign* *PACKAGE_FILE ...*

Delete all IMA and fsverity file signatures from each package
*PACKAGE_FILE* given.

## SIGN OPTIONS

*--rpmv3*

	Request RPM V3 header+payload signature addition on V4 packages.
	These signatures are expensive
	and redundant baggage on packages where a separate payload digest
	exists (packages built with rpm >= 4.14). Rpmsign will automatically
	detect the need for V3 signatures, but this option can be used to
	request their creation if the packages must be fully signature
	verifiable with rpm < 4.14 or other interoperability reasons.

	Has no effect when signing V6 packages.

*--rpmv4*

	Request RPM V4 header signature addition on V6 packages.
	Useful for making V6 packages signature verifiable
	with rpm 4.x versions.

	V4 compatibility signatures are only ever added if the signing algorithm
	is one of those known to V4: RSA, EcDSA, EdDSA (and original DSA).
	Only one V4 signature can be present in a package, so this is
	added only on the first *--addsign* with a V4 compatible
	algorithm, and ignored otherwise.

	Has no effect when signing V4 packages.

*--rpmv6*

	Request RPM V6 header signature addition on V4 packages.

	This generally always succeeds as there can be arbitrary number of
	V6 signatures on a package. A V3/V4 compatibility signatures are
	added usign the same logic as *--rpmv4* on a V6 package.

	Has no effect when signing V6 packages.

*--fskpath* *KEY*

	Used with *--signfiles*, use file signing key *KEY*.

*--certpath* *CERT*

	Used with *--signverity*, use file signing certificate *CERT*.

*--verityalgo* *ALG*

	Used with *--signverity*, to specify the signing algorithm.
	sha256 and sha512 are supported, with sha256 being the default if
	this argument is not specified. This can also be specified with the
	macro *%\_verity_algorithm*.

*--signfiles*

	Sign package files. The macro *%\_binary_filedigest_algorithm*
	must be set to a supported algorithm before building the package.
	The supported algorithms are SHA1, SHA256, SHA384, and SHA512, which
	are represented as 2, 8, 9, and 10 respectively. The file signing
	key (RSA private key) must be set before signing the package, it can
	be configured on the command line with *--fskpath* or the macro
	%\_file_signing_key.

*--signverity*

	Sign package files with fsverity signatures. The file signing key
	(RSA private key) and the signing certificate must be set before
	signing the package. The key can be configured on the command line
	with *--fskpath* or the macro %\_file_signing_key, and the cert
	can be configured on the command line with *--certpath* or the
	macro %\_file_signing_cert.


# CONFIGURING SIGNING KEYS

In order to sign packages, you need to create your own OpenPGP key pair
(aka certificate) and configure *rpm*(8) to use it. The following macros are
available:

*%\_openpgp_sign_id*

	The fingerprint or keyid of the signing key to use. Typically
	this is the only configuration needed. If omitted,
	*--key-id* must be explicitly specified when signing.

*%\_openpgp_sign*

	The OpenPGP implementation to use for signing. Supported values are
	"gpg" for GnuPG (default and traditional) and "sq" for Sequoia PGP.

Implementation specific macros:

*%\_gpg_path*

	The location of your GnuPG keyring if not the default *$GNUPGHOME*.

*%\_gpg_name*

	Legacy macro for configuring user id with GnuPG. Use the implementation
	independent and non-ambiguous *%\_openpgp_sign_id* instead.

*%\_sq_path*

	The location of your Sequoia configuration if not the default.

For example, to configure rpm to sign with Sequoia PGP using a key with
fingerprint of 7B36C3EE0CCE86EDBC3EFF2685B274E29F798E08 you would include

```
%\_openpgp_sign sq
%\_openpgp_signer 7B36C3EE0CCE86EDBC3EFF2685B274E29F798E08
```

in a macro configuration file, typically ~/.config/rpm/macros.
See *Macro Configuration* in *rpm*(8) for more details.

# SEE ALSO

*popt*(3), *rpm*(8), *rpmdb*(8), *rpmkeys*(8), *rpm2cpio*(8),
*rpmbuild*(8), *rpmspec*(8)

*rpmsign --help* - as rpm supports customizing the options via popt
aliases it's impossible to guarantee that what's described in the
manual matches what's available.

*http://www.rpm.org/ <URL:http://www.rpm.org/>*


tools/rpmsign.c
#include "system.h"
#include <errno.h>
#include <sys/wait.h>
#include <termios.h>

#include <popt.h>
#include <rpm/rpmcli.h>
#include <rpm/rpmsign.h>
#include "cliutils.h"
#include "debug.h"

enum modes {
    MODE_NONE    = 0,
    MODE_ADDSIGN = (1 << 0),
    MODE_RESIGN  = (1 << 1),
    MODE_DELSIGN = (1 << 2),
    MODE_DELFILESIGN = (1 << 3),
};

static int mode = MODE_NONE;

#if defined(WITH_IMAEVM) || defined(WITH_FSVERITY)
static int fskpass = 0;
static char * fileSigningKey = NULL;
#endif
#ifdef WITH_FSVERITY
static char * fileSigningCert = NULL;
static char * verityAlgorithm = NULL;
#endif

static struct rpmSignArgs sargs = {NULL, PGPHASHALGO_NONE, 0};

static struct poptOption signOptsTable[] = {
    { "addsign", '\0', (POPT_ARG_VAL|POPT_ARGFLAG_OR), &mode, MODE_ADDSIGN,
	N_("sign package(s)"), NULL },
    { "resign", '\0', (POPT_ARG_VAL|POPT_ARGFLAG_OR), &mode, MODE_RESIGN,
	N_("sign package(s) (identical to --addsign)"), NULL },
    { "delsign", '\0', (POPT_ARG_VAL|POPT_ARGFLAG_OR), &mode, MODE_DELSIGN,
	N_("delete package signatures"), NULL },
#if defined(WITH_IMAEVM) || defined(WITH_FSVERITY)
    { "delfilesign", '\0', (POPT_ARG_VAL|POPT_ARGFLAG_OR), &mode,
      MODE_DELFILESIGN,	N_("delete IMA and fsverity file signatures"), NULL },
#endif
    { "rpmv3", '\0', (POPT_ARG_VAL|POPT_ARGFLAG_OR),
	&sargs.signflags, RPMSIGN_FLAG_RPMV3,
	N_("create rpm v3 header+payload signatures") },
#ifdef WITH_IMAEVM
    { "signfiles", '\0', (POPT_ARG_VAL|POPT_ARGFLAG_OR),
	&sargs.signflags, RPMSIGN_FLAG_IMA,
	N_("sign package(s) files"), NULL},
#endif
#ifdef WITH_FSVERITY
    { "signverity", '\0', (POPT_ARG_VAL|POPT_ARGFLAG_OR),
	&sargs.signflags, RPMSIGN_FLAG_FSVERITY,
	N_("generate fsverity signatures for package(s) files"), NULL},
    { "verityalgo", '\0', POPT_ARG_STRING, &verityAlgorithm, 0,
	N_("algorithm to use for verity signatures, default sha256"),
	N_("<algorithm>") },
    { "certpath", '\0', POPT_ARG_STRING, &fileSigningCert, 0,
	N_("use file signing cert <cert>"),
	N_("<cert>") },
#endif
#if defined(WITH_IMAEVM) || defined(WITH_FSVERITY)
    { "fskpath", '\0', POPT_ARG_STRING, &fileSigningKey, 0,
	N_("use file signing key <key>"),
	N_("<key>") },
    { "fskpass", '\0', POPT_ARG_NONE, &fskpass, 0,
	N_("prompt for file signing key password"), NULL},
#endif
    POPT_TABLEEND
};

static struct poptOption optionsTable[] = {
    { NULL, '\0', POPT_ARG_INCLUDE_TABLE, signOptsTable, 0,
	N_("Signature options:"), NULL },
    { NULL, '\0', POPT_ARG_INCLUDE_TABLE, rpmcliAllPoptTable, 0,
	N_("Common options for all rpm modes and executables:"), NULL },

    POPT_AUTOALIAS
    POPT_AUTOHELP
    POPT_TABLEEND
};

#if defined(WITH_IMAEVM) || defined(WITH_FSVERITY)
static int flags_sign_files(int flags)
{
	return (flags & (RPMSIGN_FLAG_IMA | RPMSIGN_FLAG_FSVERITY) ? 1 : 0);
}

static char *get_fskpass(void)
{
    struct termios flags, tmp_flags;
    int passlen = 64;
    char *password = (char *)xmalloc(passlen);
    char *pwd = NULL;

    tcgetattr(fileno(stdin), &flags);
    tmp_flags = flags;
    tmp_flags.c_lflag &= ~ECHO;
    tmp_flags.c_lflag |= ECHONL;

    if (tcsetattr(fileno(stdin), TCSANOW, &tmp_flags) != 0) {
	perror("tcsetattr");
	goto exit;
    }

    printf("PEM password: ");
    pwd = fgets(password, passlen, stdin);

    if (tcsetattr(fileno(stdin), TCSANOW, &flags) != 0) {
	perror("tcsetattr");
	pwd = NULL;
	goto exit;
    }

exit:
    if (pwd)
	pwd[strlen(pwd) - 1] = '\0';  /* remove newline */
    else
	free(password);
    return pwd;
}
#endif

/* TODO: permit overriding macro setup on the command line */
static int doSign(poptContext optCon, struct rpmSignArgs *sargs)
{
    int rc = EXIT_FAILURE;
    char * name = rpmExpand("%{?_gpg_name}", NULL);

    if (rstreq(name, "")) {
	fprintf(stderr, _("You must set \"%%_gpg_name\" in your macro file\n"));
	goto exit;
    }

#if defined(WITH_IMAEVM) || defined(WITH_FSVERITY)
    if (fileSigningKey) {
	rpmPushMacro(NULL, "_file_signing_key", NULL, fileSigningKey, RMIL_GLOBAL);
    }

#ifdef WITH_FSVERITY
    if (fileSigningCert) {
	rpmPushMacro(NULL, "_file_signing_cert", NULL, fileSigningCert, RMIL_GLOBAL);
    }
    if (verityAlgorithm) {
	rpmPushMacro(NULL, "_verity_algorithm", NULL, verityAlgorithm, RMIL_GLOBAL);
    }
#endif

    if (flags_sign_files(sargs->signflags)) {
	char *fileSigningKeyPassword = NULL;
	char *key = rpmExpand("%{?_file_signing_key}", NULL);
	if (rstreq(key, "")) {
	    fprintf(stderr, _("You must set \"%%_file_signing_key\" in your macro file or on the command line with --fskpath\n"));
	    goto exit;
	}

	if (fskpass) {
	    fileSigningKeyPassword = get_fskpass();
	}

	if (fileSigningKeyPassword) {
	    rpmPushMacro(NULL, "_file_signing_key_password", NULL,
			fileSigningKeyPassword, RMIL_CMDLINE);
	    memset(fileSigningKeyPassword, 0, strlen(fileSigningKeyPassword));
	    free(fileSigningKeyPassword);
	}

	free(key);
    }
#endif

    const char *arg;
    rc = 0;
    while ((arg = poptGetArg(optCon)) != NULL) {
	if (rpmPkgSign(arg, sargs) < 0)
	    rc++;
    }

exit:
    free(name);
    return rc;
}

int main(int argc, char *argv[])
{
    int ec = EXIT_FAILURE;
    poptContext optCon = NULL;
    const char *arg;
    
    optCon = rpmcliInit(argc, argv, optionsTable);

    if (argc <= 1) {
	printUsage(optCon, stderr, 0);
	goto exit;
    }

    if (poptPeekArg(optCon) == NULL) {
	argerror(_("no arguments given"));
    }

#if defined(WITH_IMAEVM) || defined(WITH_FSVERITY)
    if (fileSigningKey && !(flags_sign_files(sargs.signflags))) {
	argerror(_("--fskpath may only be specified when signing files"));
    }
#endif

    switch (mode) {
    case MODE_ADDSIGN:
    case MODE_RESIGN:
	ec = doSign(optCon, &sargs);
	break;
    case MODE_DELSIGN:
	ec = 0;
	while ((arg = poptGetArg(optCon)) != NULL) {
	    if (rpmPkgDelSign(arg, &sargs) < 0)
		ec++;
	}
	break;
    case MODE_DELFILESIGN:
	ec = 0;
	while ((arg = poptGetArg(optCon)) != NULL) {
	    if (rpmPkgDelFileSign(arg, &sargs) < 0)
		ec++;
	}
	break;
    case MODE_NONE:
	printUsage(optCon, stderr, 0);
	break;
    default:
	argerror(_("only one major mode may be specified"));
	break;
    }

exit:
    rpmcliFini(optCon);
    return RETVAL(ec);
}
